[now-ruby] Add `@now/ruby` Builder

nathancahill opened the pull request:
Based on the excellent work by @Ehbraheem. Has tests, including for vendored and non-vendored ruby gems. Supports rack-based servers as well as WEBrick-based procs and classes.

### Function example:

`index.rb`
```ruby
require 'json'

Handler = Proc.new do |req, res|
  obj = JSON.parse(req.body)

  res.body = {
    :hello => obj["name"],
  }.to_json
end
```

`now.json`
```json
"builds": [{ "src": "index.rb", "use": "@now/ruby" }]
```

### Class example:

`index-class.rb`
```ruby
require 'json'
require 'webrick'

class Handler < WEBrick::HTTPServlet::AbstractServlet
  def do_GET req, res
    res.body =  {
      :hello => "world",
    }.to_json
  end

  def do_POST req, res
    obj = JSON.parse(req.body)

    res.body =  {
      :hello => obj["name"],
    }.to_json
  end
end
```
`now.json`
```json
"builds": [{ "src": "index-class.rb", "use": "@now/ruby" }]
```

###  Sinatra example:

`app.rb`
```ruby
require 'sinatra'
require 'json'

before do
  content_type :json
end

post '/' do
  obj = JSON.parse(request.body.read)
  { hello: obj["name"] }
end

after do
  response.body = JSON.dump(response.body)
end
```
`index.ru`
```ruby
require './app'

run Sinatra::Application
```
`now.json`
```json
"builds": [{ "src": "index.ru", "use": "@now/ruby" }]
```

### Rails example

Rails works out of the box:

`now.json`
```json
"builds": [{ "src": "app/config.ru", "use": "@now/ruby" }]
```

nathancahill commented:
I believe the CircleCI build needs to be run by someone in the Zeit organization for the build to pass with Zeit credentials.

styfle commented on packages/now-ruby/gem-install.js:
```diff
@@ -0,0 +1,71 @@
+const build_utils_1 = require('@now/build-utils');
+// portable ruby from https://portableruby.com/ for bootstrapping
+// ruby deployment on nodejs8.10. version 2.5.3 matches ruby2.5.
+const url = 'https://portableruby.s3.amazonaws.com/ruby-2.5.3.zip';
```
This looks like you own portableruby. Is there an official ruby install we could use instead?

styfle commented on packages/now-ruby/gem-install.ts:
```diff
@@ -0,0 +1,70 @@
+
+// portable ruby from https://portableruby.com/ for bootstrapping
+// ruby deployment on nodejs8.10. version 2.5.3 matches ruby2.5.
+const url = 'https://portableruby.s3.amazonaws.com/ruby-2.5.3.zip';
```
This looks like you own portableruby. Is there an official ruby install we could use instead?

nathancahill replied:
There is no official portable build, no.

styfle commented on packages/now-ruby/gem-install.js:
```diff
@@ -0,0 +1,71 @@
+'use strict';
```
Can you add `.js` to the `.gitignore` and remove these files? We only need to commit the .ts` sources.

styfle commented on packages/now-ruby/test/fixtures/01-cowsay/index.rb:
```diff
@@ -0,0 +1,10 @@
+require 'webrick'
+require 'cowsay'
+
+class Handler < WEBrick::HTTPServlet::AbstractServlet
```
It looks like `WEBrick` is a 3rd party package.

Can we instead use a core package that is shipped with Ruby?

This will avoid breaking user's code if we have to update `WEBrick` at some point.

nathancahill replied:
WEBrick is part of the Ruby stdlib, just has a funky name: https://ruby-doc.org/stdlib-2.0.0/libdoc/webrick/rdoc/WEBrick/HTTPServer.html

styfle replied:
Does this Builder support `rails` and other popular frameworks?

Ehbraheem replied:
What is the purpose of `cowsay` ?

styfle replied:
@Ehbraheem all of our tests start with 01-cowsay to make sure dependencies for the given language works properly. It emits plain text with a cow. https://en.wikipedia.org/wiki/Cowsay

styfle commented on packages/now-ruby/test/fixtures/02-cowsay-vendored/vendor/bundle/ruby/2.5.0/bin/cowsay:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env ruby
```
Can you add the `vendor` folder to `gitignore`?

styfle commented:
This is awesome! I would like @Ehbraheem to review this PR too.

@nathancahill I invited your to the ZEIT Slack workspace so we can collaborate more closely 👍

styfle commented on packages/now-ruby/now_init.rb:
```diff
@@ -0,0 +1,88 @@
+  body = payload['body']
+
+  if not SupportedHTTPMethods.key?(httpMethod)
+    return { :statusCode => 405, :body => 'Method Not Allowed' }
```
Is there a reason for performing this check manually? What will happen if you didn't check?

I ask because we have a routing layer already that performs some validation and we could probably leave it up to the user to decide which HTTP Methods they want to allow.

nathancahill replied:
Yep, doing it since Ruby uses enums for http methods. Consolidating this comment below.

styfle commented on packages/now-ruby/now_init.rb:
```diff
@@ -0,0 +1,88 @@
+  end
+
+  req = HTTParty::Request.new(
+    SupportedHTTPMethods[httpMethod],
```
Oh I guess this is where you need the method mapping. That makes sense.

styfle replied:
But why httparty? This one looks like a 3rd party dependency.

nathancahill replied:
Yes, it is 3rd party. It's similar to `request` in node. I'm not sure if there's a simple way in the stdlib to make http requests but I'll investigate.

coetry replied:
@nathancahill I would really encourage cutting out third party libs as much as possible. Can you elaborate on what you mean by arbitrary methods?

https://docs.ruby-lang.org/en/2.0.0/Net/HTTP.html

coetry replied:
@nathancahill you and send methods to Ruby classes dynamically via the `.send` method. 

```ruby
require 'net/http'
Net::HTTP.send("get", URI('https://github.com'))
```

coetry replied:
Here's a nice article that touches on metaprogramming in Ruby: https://azemoh.com/2017/01/31/metaprogramming-invoking-methods/

nathancahill replied:
@coetry Maybe you can take a look at it as it stands currently. Not sure why I'm not getting a response back with `'net/http'`

lambda responses with `socket hang up`

nathancahill replied:
Fixed this. Ruby's status code in the response is a string, lambda wanted an int 🤷‍♂️

styfle commented on packages/now-ruby/index.ts:
```diff
@@ -0,0 +1,228 @@
+import mm from 'micromatch';
+const readFile = promisify(fs.readFile);
+const writeFile = promisify(fs.writeFile);
+const stat = promisify(fs.stat);
```
I think you can remove `import fs from 'fs'` and instead import these promisified from `fs-extra`.

styfle commented on packages/now-ruby/index.ts:
```diff
@@ -0,0 +1,228 @@
+import { join, dirname } from 'path';
+import { promisify } from 'util';
+import execa from 'execa';
+import { ensureDir, move, remove } from 'fs-extra';
```
```suggestion
import { ensureDir, move, remove, stat, readFile, writeFile } from 'fs-extra';
```

styfle commented on packages/now-ruby/index.ts:
```diff
@@ -0,0 +1,228 @@
+
+  // static analysis is impossible with ruby.
+  // instead, provide `include` and `exclude` config options to reduce bundle size.
+  if (config && (config.include || config.exclude)) {
```
This should be renamed `includeFiles` and  to match the other builders. We don't have exclude yet so I think it should be removed.

styfle commented on packages/now-ruby/index.ts:
```diff
@@ -0,0 +1,228 @@
+  } catch (e) {}
+
+  console.log('calculating size of vendor directory...');
+  await execa('du', ['-hs', vendorDir], { stdio: 'inherit' });
```
I think you can remove `du` here

styfle commented on packages/now-ruby/index.ts:
```diff
@@ -0,0 +1,228 @@
+
+  let outputFiles = await glob('**', workPath);
+
+  // static analysis is impossible with ruby.
```
It looks like ruby added experimental AST support recently. Probably could use this once it goes stable. https://blog.bigbinary.com/2018/10/02/ruby-2-6-adds-rubyvm-ast-module.html

styfle commented on packages/now-ruby/gem-install.ts:
```diff
@@ -0,0 +1,70 @@
+  console.log('downloading "ruby-2.5.3.zip"...');
+  const res = await fetch(url);
+
+  if (!res.ok || res.status !== 200) {
```
```suggestion
  if (res.status !== 200) {
```

You don't need to check both because [ok](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok) is 200-299.

styfle commented on packages/now-ruby/index.ts:
```diff
@@ -0,0 +1,223 @@
+import { join, dirname } from 'path';
+import execa from 'execa';
+import { ensureDir, move, remove, stat, readFile, writeFile } from 'fs-extra';
+import mm from 'micromatch';
```
```suggestion
import { glob } from '@now/build-utils';
```

Let's use glob instead since `includeFiles` can be a glob

styfle replied:
Actually you can remove this line and the `micromatch` dependency since glob is imported below.

styfle replied:
@nathancahill Can `micromatch` be removed?

styfle commented:
@nathancahill Does this Builder support Rails, Sinatra, and other popular frameworks?

I think that support for [rack](https://rack.github.io/) should be sufficient. Then add a test for Rails and Sinatra.

nathancahill commented:
Adding those right now. Stay tuned

Ehbraheem commented:
@styfle One of the PR that you closed already support Rack based apps except Rails. [SInatra](https://sinatra.ehbraheem.now.sh)

I am still finding typescript funny. I will look at getting this up to date with rack.

Ehbraheem commented:
@nathancahill I just tested this branch with one of my rack example. It fails, you can see the logs here [https://rack-47qy43tmq.now.sh/_logs](https://rack-47qy43tmq.now.sh/_logs) it is a public deployment so you can see this source.

This is a working Rack example [https://rack-hbcfiz8oi.now.sh/](https://rack-hbcfiz8oi.now.sh/)

Ehbraheem commented:
@nathancahill Let me know if you'd need some inputs.

nathancahill commented:
Updated with rack support and tests for sinatra and rails

nathancahill commented:
Reworked the PR with out of the box Rails support by installing `ruby-devel` headers in the Fargate image. Will probably need those packages added to the base image for speed. Here's the full list (not sure which ones are installed already): https://github.com/zeit/now-builders/pull/454/files#diff-c84f2f276ff586b02c61fdb04a5dc270R15

Updated the top-level description on the PR with an example for Sinatra.

nathancahill commented:
Docs for `@now/ruby` modified from `@now/python`:

nathancahill commented:
# Ruby Builder (@now/ruby)

This builder takes in a Ruby program that defines a singular HTTP handler and outputs it as a [Lambda](https://zeit.co/docs/v2/deployments/concepts/lambdas/).

## When to Use It

Whenever you want to expose an API or a function written in Ruby.

## How to Use It

Define a `index.rb` file inside a folder as follows:


```ruby
require 'cowsay'

Handler = Proc.new do |req, res|
    res.status = 200
    res['Content-Type'] = 'text/plain'
    res.body = Cowsay.say('gem:RANDOMNESS_PLACEHOLDER', 'cow')
end
```

Inside `Gemfile` define:

```ruby
source "https://rubygems.org"

gem "cowsay", "~> 0.3.0"
```

And define a `now.json` like:

```json
{
  "version": 2,
  "builds": [{ "src": "*.rb", "use": "@now/ruby" }]
}
```

## Rack Interface

Many Ruby frameworks interface with [Rack](https://rack.github.io/) for forwarding
HTTP requests to web applications written in Ruby. You can use Rack with frameworks
such as Sinatra or Rails.

Instead of defining a handler, write the application as usual, like this Sinatra application:

```ruby
require 'sinatra'

get '/*' do
  'Hello world'
end
```

Inside `Gemfile` define:

```ruby
source "https://rubygems.org"

gem "sinatra", "~> 2.0"
gem "rack", "~> 2.0"
```

Add a Rack config file at `index.ru` to start the application:

```ruby
require './app'

run Sinatra::Application
```

And define a `now.json` like:

```json
{
  "version": 2,
  "builds": [{ "src": "index.ru", "use": "@now/ruby" }]
}
```

Most frameworks use their own implementation of routing. However, you can use a catch-all route to circumvent the framework and instead use [Now Routing Layer](https://zeit.co/docs/v2/deployments/routes/) to match a route to a Lambda.

# Technical Details

## Entrypoint

The entrypoint file must be a `.rb` source file with one of the following variables defined:

- `Handler` proc that matches the `do |req, res|` signature
- `Handler` class that inherits from the `WEBrick::HTTPServlet::AbstractServlet` class

Alternatively, a `.ru` Rack config file will serve the Ruby application it contains.

## Version

Ruby 2.5.3 is used.

## Dependencies

This builder supports installing dependencies defined in the `Gemfile`. Alternatively,
dependencies can be vendored in the root directory of the project (useful for gems that require native extentions).

## Maximum Lambda Bundle Size

To help keep cold boot times low, the [maximum output bundle size](https://zeit.co/docs/v2/deployments/concepts/lambdas/#maximum-bundle-size) for a Ruby lambda is, by default, 5mb. This [limit is extendable](https://zeit.co/docs/v2/deployments/builders/overview/#configuring-output-lambda-size) up to 50mb.

Rails, for example, requires `15mb` for an empty build.

Example `maxLambdaSize` configuration:

```json
{
  "builds": [
    { "src": "*.rb", "use": "@now/ruby", "config": { "maxLambdaSize": "15mb" } }
  ]
}
```

styfle commented:
Amazing! 💯 

It looks like the tests are failing so maybe it was just a timeout.
Running again to see what the error message is.

styfle commented on packages/now-ruby/tsconfig.json:
```diff
@@ -0,0 +1,18 @@
+      "noFallthroughCasesInSwitch": true,
+      "noImplicitReturns": true,
+      "noUnusedLocals": true,
+      "noUnusedParameters": true,
```
```suggestion
      "noUnusedParameters": true,
      "outDir": "dist",
```

styfle commented on packages/now-ruby/package.json:
```diff
@@ -0,0 +1,24 @@
+  "dependencies": {
+    "execa": "^1.0.0",
+    "fs-extra": "^7.0.1",
+    "micromatch": "^4.0.2"
```
Please remove `micromatch`. I don't think its necessary.
Also you have it in both `dependencies` and `devDependencies`

styfle commented on packages/now-ruby/.gitignore:
```diff
@@ -0,0 +1 @@
+*.js
```
```suggestion
/dist

```

styfle commented on .eslintignore:
```diff
@@ -12,3 +12,4 @@
 /packages/now-optipng/dist/*
 /packages/now-go/*
 /packages/now-rust/dist/*
+/packages/now-ruby/*
```
```suggestion
/packages/now-ruby/dist/*
```

styfle commented on packages/now-ruby/.npmignore:
```diff
@@ -0,0 +1,5 @@
+test
+tsconfig.json
+package-lock.json
+yarn.lock
```
You can delete this file and add `files` key to `package.json` instead

styfle commented on packages/now-ruby/package.json:
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "@now/ruby",
+  "author": "Nathan Cahill <nathan@nathancahill.com>",
+  "version": "0.1.0-canary.2",
```
```suggestion
  "version": "0.1.0-canary.2",
  "files": [
    "dist",
    "now_init.rb"
  ],
```

styfle commented on packages/now-ruby/index.ts:
```diff
@@ -0,0 +1,203 @@
+    await remove(join(vendorDir, 'cache'));
+  } catch (e) {}
+
+  const originalRbPath = join(__dirname, 'now_init.rb');
```
```suggestion
  const originalRbPath = join(__dirname, '..', 'now_init.rb');
```

styfle commented on packages/now-ruby/test/fixtures/05-sinatra/index.ru:
```diff
@@ -0,0 +1,3 @@
+require './app'
+
+run Sinatra::Application
```
Is this `index.ru` file necessary? What's the difference between `.ru` and `.rb`?

nathancahill replied:
Yes, this is the standard for serving Rack applications. See the Technical Details section here: https://github.com/zeit/now-builders/pull/454#issuecomment-497489011

styfle replied:
Great, so it looks like Rack supports both `.ru` and `.rb` according to [their wiki tutorial](https://github.com/rack/rack/wiki/(tutorial)-rackup-howto).

styfle commented on packages/now-ruby/test/fixtures/06-rails/app/Gemfile:
```diff
@@ -0,0 +1,28 @@
+source 'https://rubygems.org'
```
The file above `packages/now-ruby/test/fixtures/06-rails/app/.ruby-version` appears to be empty. Is that expected?

styfle commented on packages/now-ruby/test/fixtures/06-rails/app/README.md:
```diff
@@ -0,0 +1,24 @@
+# README
```
I think you can delete this file

styfle commented on packages/now-ruby/test/fixtures/06-rails/app/app/assets/stylesheets/application.css:
```diff
@@ -0,0 +1,15 @@
+/*
```
I think you can delete this file

styfle commented on packages/now-ruby/test/fixtures/06-rails/app/config/credentials.yml.enc:
```diff
@@ -0,0 +1 @@
+mlc49SVpvRzvhQHNstg/DYms2QRSFBOzHcHaazRXSQfQcMdWHyjKKHsvWqilErM5/n/jy4gUz8HaRYiFt9N1AU7nAXCinTD6d3C+5XaqkHyIwmvmrIahxHqu9xrZ4koS3pVRbCW0tKyWCwDupD6Id9N5HwxTuM5MLcHC7vBk6l+p8iRKGDou5kJe3ucAwsFwmM4xlaQwK3k/09FpUvJ4LCwmFnert2HXmwsopoEVDVHH0FJx8qP9admhkEBRPdwoK6CthQUVfaeKZZ+vo4W22zQE8h46m7Jz42p6qsV1GgpKuQJYn6gUhdwXw+cVaon4ouaPqjrOptFtVZekFpSN1raFujEaODehJevn0b0/OIWyekuCYFJB5loJZbTOoANPx7u4Ik7BQ6U9Ksd ...
```
Is this file necessary?

styfle commented on packages/now-ruby/test/fixtures/06-rails/app/config/environments/development.rb:
```diff
@@ -0,0 +1,40 @@
+Rails.application.configure do
```
I don't thik we need dev vs prod in a test fixture. Perhaps we could use the prod configuration and delete this one?

styfle commented on packages/now-ruby/test/fixtures/06-rails/app/config/initializers/application_controller_renderer.rb:
```diff
@@ -0,0 +1,8 @@
+# Be sure to restart your server when you modify this file.
```
Can the files that are all comments be deleted?

styfle commented on packages/now-ruby/test/fixtures/06-rails/app/config/locales/en.yml:
```diff
@@ -0,0 +1,33 @@
+# available at http://guides.rubyonrails.org/i18n.html.
+
+en:
+  hello: "Hello world"
```
Can locals be deleted?

styfle commented on packages/now-ruby/test/fixtures/06-rails/app/config/master.key:
```diff
@@ -0,0 +1 @@
+be4e97f95c0cbbf9e57a075cfb672e9f
```
What is this? Can it be deleted? 😄

nathancahill replied:
Rails needs a `master.key` and a `credentials.yml.enc` file to run.

styfle commented on packages/now-ruby/test/fixtures/06-rails/now.json:
```diff
@@ -0,0 +1,16 @@
+  "probes": [{ "path": "/", "mustContain": "gem:RANDOMNESS_PLACEHOLDER" }],
+  "env": {
+    "RAILS_ENV": "production",
+    "RAILS_LOG_TO_STDOUT": "1"
```
Let's remove these `env` variables and assign automatically in the Builder.

Something like this should do it:


```js
if (typeof process.env.RAILS_ENV === 'undefined') {
  process.env.RAILS_ENV = 'production';
}
if (typeof process.env.RAILS_LOG_TO_STDOUT === 'undefined') {
  process.env.RAILS_LOG_TO_STDOUT = '1';
}
```

This will help new Ruby users get started easily 👍

styfle commented on packages/now-ruby/test/test.js:
```diff
@@ -0,0 +1,33 @@
+  testDeployment,
+} = require('../../../test/lib/deployment/test-deployment.js');
+
+jest.setTimeout(4 * 60 * 1000);
```
```suggestion
jest.setTimeout(5 * 60 * 1000);
```

styfle commented on yarn.lock:
```diff
@@ -6983,6 +7014,14 @@ micromatch@^3.1.10, micromatch@^3.1.4, micromatch@^3.1.8:
     snapdragon "^0.8.1"
     to-regex "^3.0.2"
 
+micromatch@^4.0.2:
```
I thought you removed `micromatch`?

I looks like you're only using `execa` and `fs-extra` so perhaps you could revert the `yarn.lock` file and then remove `dependencies` key from package.json.

Then you can run `yarn add execa@1.0.0 fs-extra@7.0.1` and I would expect it to install without modifying the `yarn.lock` since those already exist in the workspace.

nathancahill replied:
Thanks!

styfle commented on packages/now-ruby/index.ts:
```diff
@@ -0,0 +1,209 @@
+
+const REQUIRED_VENDOR_DIR = 'vendor/bundle/ruby/2.5.0';
+
+async function dirExists(directory: string) {
```
You can use `pathExists` from `fs-extra`

styfle commented on packages/now-ruby/index.ts:
```diff
@@ -0,0 +1,226 @@
+
+interface RubyConfig extends Config {
+  excludeFiles?: string | string[];
+}
```
You can undo this change now that `Config` has the proper type `[key: string]: string | string[] | boolean | number | undefined;`

styfle commented:
Published as `@now/ruby@canary` just now.
